src/main/java/sparksoniq/spark/closures/ClosureUtils.java

```diff
@@ -30,6 +31,21 @@
         return byteArray;
     }
 
+    public static Row reserializeRowWithNewData(Row prevRow, List<Item> newColumn, int duplicateColumnIndex) {
+        List<byte[]> newRowColumns = new ArrayList<>();
```

Comment:
This should also be a field of the closure class, constructed once there, and passed as a parameter to this function. It must be cleared every time.

A few more thoughts:
The input and output buffers should also be fields in the closure class, as well as the input and output objects, using these buffers. The use of array byte streams is documented as superfluous by kryo. The function serializeItemList should then return Output.position() rather than the buffer (there is a "memory leak" here, as a lot of byte arrays are created, which are longer than the bytes actually used). Then, only the relevant part of the buffer (0 to returned position) should be copied over into the row to optimize space.

Reply:
I have briefly experimented with the Output.toBytes(). This works with the output part of the picture but when the data is read through the input, following exception is thrown:
org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 15.0 failed 1 times, most recent failure: Lost task 1.0 in stage 15.0 (TID 61, localhost, executor driver): com.esotericsoftware.kryo.KryoException: Buffer underflow.

I will play around a bit more with the Kryo input/output classes.

Reply:
Thanks! If this causes issues, feel free to merge the PR and leave this aside for later. This is something that I think is not in the way of starting with the change to Spark SQL, i.e., it can be investigated in parallel (this code will be in a user-defined function rather than a closure, but the spirit remains identical).

Reply:
I have managed to get the Output.toBytes() working.

Additionally, ByteBufferInput/Output caught my eye in the kryo documentation and I have started using these as our input/output classes. They have provided a somewhat noticeable performance boost in our local tests. I'm hoping their effects will show up in the tests with data.

Reply:
Very nice!

Also good to hear about these alternative input/outputs. The performance boost sounds good.